### ISSUE TITLE
Captured log to outfile

### ISSUE THREAD

**Issue by janoskut** [behave/behave#979]
There doesn't seem to be a way to get the captured logs into a log file, using the `--outfile` option.

I can get it to the console using `--no-logcapture` and `--format=plain`, but not to the output file using any formatter.

How would I go ahead if I wanted:
- `--format=pretty` in the console (no log output on success)
- `--format=plain` with `--logging-level=DEBUG` to an `--outfile` output file (on success and failure, no difference)

?

Related to: https://github.com/behave/behave/issues/412

**Comment by jenisys** [comment]
The `--outfile` option is used to redirect the output of a formatter to a file.
This means:

* You have a primary formatter. It uses either console/stdout or an output-file.
* You may have many (0..N) secondary formatters. Each formatter will use its own output-file.
  Secondary formatters are often defined in the `"behave.ini"` config-file.

OTHERWISE: Logging is for diagnostics.
To redirect the logging output to a file, setup the logging subsystem accordingly.
This setup is best done in the `before_all()` hook (otherwise it is too late for log-capture, etc.).

SEE ALSO:
* [features/logging.setup_with_configfile.feature](https://github.com/behave/behave/blob/master/features/logging.setup_with_configfile.feature)
* [features/runner.multiple_formatters.feature](https://github.com/behave/behave/blob/master/features/runner.multiple_formatters.feature)

**Comment by janoskut** [comment]
Thanks for your explanations @jenisys . I think that is exactly as far as I got.

So how would I go about putting formatter-output and logging output to the same file? I need the full output of `behave --format plain --logging-level=DEBUG --no-logcapture` in a logfile, but I don't want to have it printed to stdout. So also redirecting stdout to a file is not an option.

I tried: using the same file for `-o` and the logging file handler, but behave formatters overwrite the file. Even with an own formatter which appends to a file ('a' instead of 'w' file mode), the logging is written first into the file, and the formatter output is added afterwards. I guess this could be somehow fixed, but writing into the same file from different directions doesn't seem right to me.

Do I need to write a formatter which makes use of the captured logs at each step? How would I do that?

**Comment by jenisys** [comment]
Why would you want to do that?
* The result would garbled text.
* The Scenario/Step output with the formatter would be garbled with log-record statements

If you really want to that:
* Formatter output should use stdout
* Logging output should use stdout/console
* You redirect the output to a file (and stderr probably too)

```console
# -- ASSUMPTION: Logging handler(s) write output to stdout
$ behave -f plain features > combined_output.txt
```

**Comment by janoskut** [comment]
> Why would you want to do that?

Why would I want to do what exactly? For me it's a pretty obvious use case: I want a logfile which has behave's explanation of what steps are executed, and in between the log output of the thing under test. Just as I have when I use `plain --logging-level=DEBUG --no-logcapture`. Because it is too verbose, I want this detail in a logfile, not in the console. On the console, I'm happy with the "pretty" formatter.

I guess I could use the `before_step()` etc. hooks to keep track of what is executed in the log. It doesn't look like the formatter API is capable of knowing when a step is started, or would that be a way to write a formatter which does the logging?

**Comment by janoskut** [comment]
> Why would you want to do that?

Why would I want to do what exactly? For me it's a pretty obvious use case: I want a logfile which has behave's explanation of what steps are executed, and in between the log output of the thing under test. Just as I have when I use `plain --logging-level=DEBUG --no-logcapture`. Because it is too verbose, I want this detail in a logfile, not in the console. On the console, I'm happy with the "pretty" formatter.

I guess I could use the `before_step()` etc. hooks to keep track of what is executed in the log. It doesn't look like the formatter API is capable of knowing when a step is started, or would that be a way to write a formatter which does the logging?

**Comment by jenisys** [comment]
Mmh, how is step execution then tracked and output for it generated by a `Formatter`, especially in the `PrettyFormatter` ?!?
TLDR, the `Formatter.step()` is called when the step execution starts.

DETAILS:
* A trivial `Formatter` only reports the step after it is executed, because it needs to report the step-result, too.
* The `PrettyFormatter` reports the start of the step execution and overwrites this/these line(s) after the step-result is known.

**Comment by hpe-ykoehler** [comment]
I agree with @janoskut that it would be great if the steps were recorded in the log

Log
<timestamp> INFO:builtins.step_begin:Given a device ...
<timestamp> DEBUG:<component>:message
<timestamp> INFO:builtins.step_end:Given a device
Console:
  Given a device ... <grey/green/whatever>

The concept of displaying log only on error for diagnostics assumes that there are no false positive tests, which require the log to be investigated, and that the underlying logic used to accomplish those steps has no value.

Even when a test passes, all records proving and explaining what steps were done, etc., need to be recorded in enterprise testing.  False positives may not be detected easily in a complex environment, and when they are detected, having this data allows tracking when they were introduced and gives historical information.

Also, understanding how the test was executed is as important as the result. A Pass or fail has no value without proof, which is kind of what we learn in school for maths as well. How you got there is as important, if not more important, than the final results.

I think it would be beneficial for behave framework to enable recording the log in a file at all time, so that it can be stored along with the console output into a ticket management system to track test execution and such.

**Comment by jenisys** [comment]
@hpe-ykoehler 
Mmh, I think I answered the question already and you seem to have the wrong understanding on some parts.

* Provide a formatter that logs steps and use it (if you need that)
* Multiple formatters are already allowed and supported at the same time, even user-defined one.
* Put the default formatters that log to files into the behave config-file
* Plain and pretty formatter can already be used to record the details as test report in a file (in addition to another formatter).

LOGGING RELATED:

* If you are logging to a file, any submitted log record will appear in this file.
* This is independent from the fact that log-capture mode is used or not or any logging-filter is used or not.

**Comment by jenisys** [comment]
@janoskut
 
* Example how to set up logging to a file (see below).
* Log records to a file will always be present in the log file (as stated above).
* This is independent from if the log-capture mode is enabled or not.


SEE ALSO:

* [docs: Improved logging support](https://behave.readthedocs.io/en/latest/new_and_noteworthy_v1.2.7/#improved-logging-support)
* [features/logging.to_file.feature](https://github.com/behave/behave/blob/main/features/logging.to_file.feature)
